Add Duffing spring example (#26641)
sympy/sympy: doc/src/tutorials/physics/mechanics/duffing-example.rst

- :math:`\dot{q}_2`: First time derivative of `q2`, representing the angular velocity of the pendulum
- :math:`u_1`: Generalized speed associated with the Duffing oscillator
- :math:`u_2`: Generalized speed associated with the pendulum
- :math:`\dot{u}_1`: First time derivative of `u1`, representing the acceleration of the Duffing oscillator
- :math:`\dot{u}_2`: First time derivative of `u2`, representing the angular acceleration of the pendulum

Define Kinematics
=================

Define all the reference frames and points.

   >>> N = me.ReferenceFrame('N')
   >>> B = N.orientnew('B', 'axis', (q2, N.z))

The angular velocity of the pendulum in the reference frame is:

   >>> B.set_ang_vel(N, q2.diff() * N.z)
   >>> B.ang_vel_in(N)
   Derivative(q2(t), t)*N.z

Locations and velocities of the Duffing Oscillator block and the pendulum are:

   >>> O = me.Point('O')
   >>> Block = O.locatenew('Block', q1 * N.y)
   >>> Pendulum = Block.locatenew('Pendulum', l * B.y)

O is a fixed point in the inertial reference frame.

   >>> O.set_vel(N, 0)
   >>> Block.set_vel(N, q1d * N.y)
   >>> Pendulum.v2pt_theory(Block, N, B)

>    Derivative(q1(t), t)*N.y - l*Derivative(q2(t), t)*B.x

Reviewer: You can use `me.init_vprinting()` to have pretty printed output for these lines.

Define inertia and rigid bodies.
Here, we assume a simple pendulum which consists of a bob of mass m hanging from a massless string of length l
and fixed at a pivot point (Duffing Oscillator Block).

   >>> I_block = me.inertia(N, M*(h**2 + d**2)/12, M*(w**2 + h**2)/12, M*(w**2 + d**2)/12)
   >>> I_pendulum = me.inertia(B, 0, 0, 2*m*r**2/5)

   >>> par_block = me.RigidBody('block', Block, N, M, (I_block, Block))
   >>> par_pendulum = me.RigidBody('pendulum', Pendulum, B, m, (I_pendulum, Pendulum))

Define Force, Energy
====================

We obtain the Duffing force using the `DuffingSpring` actuator from ``sympy.physics.mechanics.actuator``.
This force will be used to calculate the potential energy of the Duffing Oscillator block.

Define the Duffing spring force.

   >>> pathway = me.LinearPathway(O, Block)
   >>> duffing_spring = me.DuffingSpring(k1, k2, pathway, 0)
   >>> duffing_force = -duffing_spring.force
   >>> duffing_force
   k1*sqrt(q1(t)**2) + k2*(q1(t)**2)**(3/2)

Define Rayleigh dissipation.

>    >>> D = (1/2) * c1 * q1d**2

Reviewer: For this term it would be ideal if we had a `LinearDamper()` acuator, then you can pass use `.to_loads()` from both to generate the force list for `LagrangesMethod`. If we don't have that yet, then it would still be best to add this a a load to `LagrangesMethod` instead of forming the generalized force manually.

Reviewer: Also, you write the energy term here as `D` but then you pass it in as a force below in the force list. `D` is not a force but the energy lost due to a linear damping force. Best to define as a force `-c1*q1d*N.y`.

.. _duffing-example:

==================================
Duffing Oscillator with a Pendulum
==================================

In this example we demonstrate the use of functionality provided in
:obj:`sympy.physics.mechanics` for deriving the quations of motion (EOM) for a system

> consisting of a Duffing oscillator with a pendulum.

Reviewer 2: Would be nice to refer that this example is inspired by that paper (I think you can use something like `:cite:...`).

.. _fig-duffing-oscillator-pendulum:
.. figure:: duffing-oscillator-pendulum.svg

The system will be modeled using Lagrange equations. `M` is mass of the Duffing oscillator,
`m` is mass of the pendulum, `l` is length of the pendulum. `k_1` and `k_2` are linear and
non-linear parts of spring stiffness, and `c_1` is a viscous damping coefficient of the Duffing oscillator.

   >>> import sympy as sm
   >>> import sympy.physics.mechanics as me

Define Variables
================

   >>> M, m, l, k1, k2, c1, g, h, w, d, r = sm.symbols('M, m, l, k1, k2, c1, g, h, w, d, r')

>    >>> q1, q2 = me.dynamicsymbols('q1 q2')
>    >>> q1d, q2d = me.dynamicsymbols('q1 q2', 1)
>    >>> u1, u2 = me.dynamicsymbols('u1 u2')
>    >>> u1d, u2d = me.dynamicsymbols('u1 u2', 1)

Reviewer 2: ```suggestion
   >>> q1, q2, u2, u2 = me.dynamicsymbols('q1 q2 u1 u2')
```
You don't use time-derivatives manually, so you can remove them.

- :math:`h`: Height of the Duffing oscillator
- :math:`w`: Width of the Duffing oscillator
- :math:`d`: Depth of the Duffing oscillator
- :math:`r`: Radius of the massive bob of the pendulum
- :math:`q_1`: Generalized coordinate representing the position of the Duffing oscillator
- :math:`q_2`: Generalized coordinate representing the angle of the pendulum
- :math:`\dot{q}_1`: First time derivative of `q1`, representing the velocity of the Duffing oscillator

> - :math:`\dot{q}_2`: First time derivative of `q2`, representing the angular velocity of the pendulum

Reviewer 2: These can be removed, they are self-explanatory.

- :math:`u_1`: Generalized speed associated with the Duffing oscillator
- :math:`u_2`: Generalized speed associated with the pendulum
- :math:`\dot{u}_1`: First time derivative of `u1`, representing the acceleration of the Duffing oscillator

> - :math:`\dot{u}_2`: First time derivative of `u2`, representing the angular acceleration of the pendulum

Reviewer 2: These can be removed, they are self-explanatory.

Define Kinematics
=================

Define all the reference frames and points.

   >>> N = me.ReferenceFrame('N')
   >>> B = N.orientnew('B', 'axis', (q2, N.z))

The angular velocity of the pendulum in the reference frame is:

>    >>> B.set_ang_vel(N, q2.diff() * N.z)

Reviewer 2: This angular velocity has already been set automatically when orieting the frame, so you can remove this line.

   >>> B.ang_vel_in(N)
   Derivative(q2(t), t)*N.z

Locations and velocities of the Duffing Oscillator block and the pendulum are:

   >>> O = me.Point('O')

>    >>> Block = O.locatenew('Block', q1 * N.y)
>    >>> Pendulum = Block.locatenew('Pendulum', l * B.y)

Reviewer 2: ```suggestion
   >>> block = O.locatenew('block', q1 * N.y)
   >>> pendulum = block.locatenew('pendulum', l * B.y)
```
Use lowercase, according to the pep8 standard as mentioned by @moorepants

O is a fixed point in the inertial reference frame.

   >>> O.set_vel(N, 0)
   >>> Block.set_vel(N, q1d * N.y)
   >>> Pendulum.v2pt_theory(Block, N, B)
   Derivative(q1(t), t)*N.y - l*Derivative(q2(t), t)*B.x

>    >>> me.init_vprinting()

Reviewer 2: You should move this line just after the imports, e.g.:
```
>>> import sympy as sm
>>> import sympy.physics.mechanics as me
>>> me.init_vprinting()
```

Define inertia and rigid bodies.
Here, we assume a simple pendulum which consists of a bob of mass m hanging from a massless string of length l
and fixed at a pivot point (Duffing Oscillator Block).

   >>> I_block = me.inertia(N, M*(h**2 + d**2)/12, M*(w**2 + h**2)/12, M*(w**2 + d**2)/12)
   >>> I_pendulum = me.inertia(B, 0, 0, 2*m*r**2/5)

   >>> par_block = me.RigidBody('block', Block, N, M, (I_block, Block))

>    >>> par_pendulum = me.RigidBody('pendulum', Pendulum, B, m, (I_pendulum, Pendulum))

Reviewer 2: I would propose just using particles as done in the paper. This allows comparing the final equations to the paper more easily and it makes the equations more readable as they are simpler.

When doing so, don't forget to remove `h`, `d`, `w`, and `r` from the start of the example.

Author: I compared the equations of motion derived from `LagrangesMethod `.
Equations of motion in the paper:
$$(M+m)q_{1}'' - mlq_{2}''sin(q_{2})-ml(q_{2}')^2cos(q_{2})+k_{1}q_{1}+k_{2}(q_{1})^3+c_{1}q_{1}'$$
$$ml^2q_{2}''-mlq_{1}''sin(q_{2})+mlgsin(q_{2})+c_{2}q_{2}'$$
Equations of motion in this example:
$$(M+m)q_{1}'' - mlq_{2}''sin(q_{2})-ml(q_{2}')^2cos(q_{2})+k_{1}q_{1}+k_{2}(q_{1})^3+c_{1}q_{1}' -(M+m)g$$
$$ml^2q_{2}''-mlq_{1}''sin(q_{2})+mlgsin(q_{2})+ \frac{2r^2q_{2}''}{5}$$

The differences come from the gravitational force and the existence of the damping torque with damping coefficient $c_2$, so I believe equations of motion obtained in this example are acceptable

Reviewer 2: Part of the reason was that it also kept the tutorial simpler (the block doesn't rotate so those inertia terms are by definition not showing up in any of the equations), but keeping it is fine with me.

Define Forces
=============

We calculate the forces acting on the system.

   >>> path = me.LinearPathway(O, Block)
   >>> spring = me.DuffingSpring(k1, k2, path, 0)
   >>> damper = me.LinearDamper(c1, path)

   >>> loads = spring.to_loads() + damper.to_loads()

   >>> bodies = [par_block, par_pendulum]

   >>> for body in bodies:

>    ...     gravitational_force = -body.mass * g * N.y
>    ...     gravitational_load = (body.masscenter, gravitational_force)
>    ...     loads.append(gravitational_load)

Reviewer 2: ```suggestion
   ...     loads.append(me.Force(body, -body.mass * g * N.y))
```

         |     /   2       /  2\   |                  |       /   2       /  2\   |
         \k1*\/  q1   + k2*\q1 /   /*q1               \- k1*\/  q1   - k2*\q1 /   /*q1
    [(O, ------------------------------ n_y), (Block, -------------------------------- n_y), (O, c1*q1'(t) n_y), (Block, -c1*q1'(t) n_y), (Block, -M*g n_y), (Pendulum, -g*m n_y)]
                       _____                                         _____
                      /   2                                         /   2
                    \/  q1                                        \/  q1

Lagrange's Method
=================

With the problem setup, the Lagrangian can be calculated, and the equations of motion formed.

   >>> L = me.Lagrangian(N, par_block, par_pendulum)
   >>> me.Lagrangian(N, par_block, par_pendulum)
            2      2       2     / 2       2                                     2\
    M*q1'(t)    m*r *q2'(t)    m*\l *q2'(t)  - 2*l*sin(q2)*q1'(t)*q2'(t) + q1'(t) /
    --------- + ------------ + ----------------------------------------------------
        2            5                                  2

>    >>> LM = me.LagrangesMethod(L, [q1, q2], forcelist = loads, frame = N)

Reviewer 2: ```suggestion
   >>> LM = me.LagrangesMethod(L, [q1, q2], bodies=bodies, forcelist=loads, frame=N)
```
Though `LangrangesMethod` doesn't make use of the bodies, it is nice to provide them anyways. Additionally, when calling a function keyword arguments are generally specified without spacing.

>    >>> LM.form_lagranges_equations()

Reviewer 2: ```suggestion
   >>> sm.simplify(LM.form_lagranges_equations())
```
Would recommend comparing results of `sm.simplify` and something like `sm.expand` to see which is more readable and looks closer to the EoMs shown in the paper (they are of course slightly different as you describe a slightly different system).

.. _duffing-example:

==================================
Duffing Oscillator with a Pendulum
==================================

In this example we demonstrate the use of functionality provided in

> :obj:`sympy.physics.mechanics` for deriving the quations of motion (EOM) for a system

Reviewer 2: ```suggestion
:obj:`sympy.physics.mechanics` for deriving the equations of motion for a system
```
You are not using the abbreviation anywhere in this tutorial.

paper [P.Brzeskia2012]_ section 2.

.. _fig-duffing-oscillator-pendulum:
.. figure:: duffing-oscillator-pendulum.svg

The system will be modeled using Lagrange equations. `M` is mass of the Duffing oscillator,
`m` is mass of the pendulum, `l` is length of the pendulum. `k_1` and `k_2` are linear and
non-linear parts of spring stiffness, and `c_1` is a viscous damping coefficient of the Duffing oscillator.

   >>> import sympy as sm
   >>> import sympy.physics.mechanics as me
   >>> me.init_vprinting()

Define Variables
================

   >>> M, m, l, k1, k2, c1, g, h, w, d, r = sm.symbols('M, m, l, k1, k2, c1, g, h, w, d, r')
   >>> q1, q2, u2, u2 = me.dynamicsymbols('q1 q2 u1 u2')
   >>> q1d = me.dynamicsymbols('q1', 1)

- :math:`h`: Height of the Duffing oscillator
- :math:`w`: Width of the Duffing oscillator
- :math:`d`: Depth of the Duffing oscillator
- :math:`r`: Radius of the massive bob of the pendulum
- :math:`q_1`: Generalized coordinate representing the position of the Duffing oscillator
- :math:`q_2`: Generalized coordinate representing the angle of the pendulum
- :math:`u_1`: Generalized speed associated with the Duffing oscillator

> - :math:`u_2`: Generalized speed associated with the pendulum

Reviewer: change the u variables to qdots, right?

Author: Yeah, I'll remove u variables

=================

Define all the reference frames and points.

   >>> N = me.ReferenceFrame('N')
   >>> B = N.orientnew('B', 'axis', (q2, N.z))

The angular velocity of the pendulum in the reference frame is:

   >>> B.ang_vel_in(N)
   q2'(t) n_z

Locations and velocities of the Duffing Oscillator block and the pendulum are:

   >>> O = me.Point('O')
   >>> block_point = O.locatenew('block', q1 * N.y)
   >>> pendulum_point = block_point.locatenew('pendulum', l * B.y)

O is a fixed point in the inertial reference frame.

   >>> O.set_vel(N, 0)
   >>> block_point.set_vel(N, q1d * N.y)
   >>> pendulum_point.v2pt_theory(block_point, N, B)
   q1'(t) n_y + -l*q2'(t) b_x

Define inertia and rigid bodies.
Here, we assume a simple pendulum which consists of a bob of mass m hanging from a massless string of length l
and fixed at a pivot point (Duffing Oscillator Block).

   >>> I_block = M / 12 * me.inertia(N, h**2 + d**2, w**2 + d**2, w**2 + h**2)

>    >>> I_pendulum = 2*m*r**2/5*me.inertia(B, 1, 1, 1)

Reviewer: ```suggestion
   >>> I_pendulum = 2*m*r**2/5*me.inertia(B, 1, 0, 1)
```

>    >>> I_block = me.inertia(N, M*(h**2 + d**2)/12, M*(w**2 + h**2)/12, M*(w**2 + d**2)/12)

Reviewer 2: ```suggestion
   >>> I_block = M / 12 * me.inertia(N, h**2 + d**2, w**2 + d**2, w**2 + h**2)
```
Inertia was incorrect. You could optionally use the `Inertia` object to immediately also specify the point w.r.t. which the inertia is defined.
```suggestion
   >>> I_block = me.Inertia.from_inertia_scalars(bock_point, N, M*(h**2 + d**2)/12, M*(w**2 + d**2)/12, M*(w**2 + h**2)/12)
```

>    >>> I_pendulum = me.inertia(B, 0, 0, 2*m*r**2/5)

Reviewer 2: ```suggestion
   >>> I_pendulum = 2*m*r**2/5*me.inertia(B, 1, 1, 1)
```
Same here.

   >>> par_block = me.RigidBody('block', block_point, N, M, (I_block, block_point))

>    >>> par_pendulum = me.RigidBody('pendulum', pendulum_point, B, m, (I_pendulum, pendulum_point))

Reviewer 2: Maybe rename to `block_body` or `block_rb` and `pendulum_...`, as you are modeling them as rigid bodies and not as particles.

=============

We calculate the forces acting on the system.

   >>> path = me.LinearPathway(O, block_point)
   >>> spring = me.DuffingSpring(k1, k2, path, 0)
   >>> damper = me.LinearDamper(c1, path)

   >>> loads = spring.to_loads() + damper.to_loads()

   >>> bodies = [block_body, pendulum_body]

   >>> for body in bodies:
   ...     loads.append(me.Force(body, body.mass * g * N.y))

   >>> loads
               /      _____           3/2\                  /        _____           3/2\
         |     /   2       /  2\   |                  |       /   2       /  2\   |
         \k1*\/  q1   + k2*\q1 /   /*q1               \- k1*\/  q1   - k2*\q1 /   /*q1
    [(O, ------------------------------ n_y), (block, -------------------------------- n_y), (O, c1*q1'(t) n_y), (block, -c1*q1'(t) n_y), (block, M*g n_y), (pendulum, g*m n_y)]
                       _____                                         _____
                      /   2                                         /   2
                    \/  q1                                        \/  q1

Lagrange's Method
=================

With the problem setup, the Lagrangian can be calculated, and the equations of motion formed.

>    >>> L = me.Lagrangian(N, block_body, pendulum_body)

Reviewer: The Lagrangian should technically include potential energy term. For this model, it is gravity and the Duffing spring. You can add them as loads, but it is probably worth stating that you are setting the potential energy to zero in the Lagrangian and including the conservative forces in the loads, which is equivalent.

>    >>> me.Lagrangian(N, block_body, pendulum_body)

Reviewer: ```suggestion
   >>> L
```

   >>> LM = me.LagrangesMethod(L, [q1, q2], bodies=bodies, forcelist=loads, frame=N)
   >>> sm.simplify(LM.form_lagranges_equations())
    [                                       /                                    2          \   /          2\   ]
    [-M*g + M*q1''(t) + c1*q1'(t) - g*m - m*\l*sin(q2)*q2''(t) + l*cos(q2)*q2'(t)  - q1''(t)/ + \k1 + k2*q1 /*q1]
    [                                                                                                           ]
    [                     /                   2                                    2        \                   ]
    [                   m*\5*g*l*sin(q2) + 5*l *q2''(t) - 5*l*sin(q2)*q1''(t) + 2*r *q2''(t)/                   ]
    [                   ---------------------------------------------------------------------                   ]
    [                                                     5                                                     ]

Equations of motion in [P.Brzeskia2012]_:

.. math::

   (M + m)q_1'' - mlq_2'' \sin(q_2) - ml(q_2')^2 \cos(q_2) + k_1 q_1 + k_2 (q_1)^3 + c_1 q_1' = F_0 \cos(\nu t)

   ml^2 q_2'' - mlq_1'' \sin(q_2) + mlg \sin(q_2) + c_2 q_2' = 0

> 

Reviewer 2: Isn't it possible to use `\ddot{q}_i` and `\dot{q}_i` here?

Reviewer 2: Additionally, you can probably replace `(q_1)^3` by `q_1^3`

Equations of motion in this example:

.. math::

   (M + m)q_1'' - mlq_2'' \sin(q_2) - ml(q_2')^2 \cos(q_2) + k_1 q_1 + k_2 q_1^3 + c_1 q_1' - (M + m)g = 0

   ml^2 q_2'' - mlq_1'' \sin(q_2) + mlg \sin(q_2) + \frac{2r^2q_2''}{5} = 0

> The differences in the equations of motion are attributed to several factors: the gravitational force, a damping torque characterized by the damping coefficient `c_2`,

Reviewer: I don't see $c_2$ in the equations. Is this a typo?

Author: $c_2$ appears in the equations of motion in the paper but not in our example because the paper introduces a damping torque, but our example does not. So it is normal that $c_2$ does not appear in line 136.

Reviewer: Ah, I read the equation and then you mention c_2 so I was wondering what that is. But I hadn't read further to the next equation.

Reviewer: Actually, that is wrong. I still don't see a c_2 in either of the equations you write.

Reviewer: Nevermind. I'm blind.

   >>> par_block = me.RigidBody('block', Block, N, M, (I_block, Block))
   >>> par_pendulum = me.RigidBody('pendulum', Pendulum, B, m, (I_pendulum, Pendulum))

Define Forces Using a Custom Actuator
=====================================

We define a custom actuator class, `DuffingDamperPathway`, to calculate the forces acting on the system.
This class combines the forces from a Duffing spring and a damper to model the dynamics of the Duffing oscillator block.
We also consider gravity as a force acting on the pendulum.

   >>> class DufPenPathway(me.ActuatorBase):
   ...     def __init__(self, P1, P2, P3, linear_stiffness, nonlinear_stiffness, damping, pendulum_mass, block_mass, g):
   ...         self.P1 = P1
   ...         self.P2 = P2
   ...         self.P3 = P3
   ...         self.k1 = linear_stiffness
   ...         self.k2 = nonlinear_stiffness
   ...         self.c1 = damping
   ...         self.m = pendulum_mass
   ...         self.M = block_mass
   ...         self.g = g
   ...
   ...         self.pathway1 = me.LinearPathway(self.P1, self.P2)
   ...         self.duffing_spring = me.DuffingSpring(self.k1, self.k2, self.pathway1, 0)

>    ...         self.pathway2 = me.LinearPathway(self.P2, self.P3)

Reviewer: Why would the pathways be different for the spring and the damper? Aren't they in parallel?

   ...
   ...     def to_loads(self):
   ...         duffing_force = self.duffing_spring.force * N.y
   ...         damper_force = -self.c1 * q1d * N.y
   ...         total_force_block = duffing_force + damper_force
   ...         loads = [(self.P2, total_force_block), (self.P3, -self.m * self.g * N.y)]

>    ...         return loads

Reviewer: Given that your GSoC is about adding actuators, why don't you just add a `LinearDamper`?

Author: Thank you for the advice, so then would it be like this?
```
class DuffingActuator(me.ActuatorBase):
    def __init__(self, P1, P2, linear_stiffness, nonlinear_stiffness, damping, pendulum_mass, block_mass, g):
        self.P1 = P1
        self.P2 = P2
        self.k1 = linear_stiffness
        self.k2 = nonlinear_stiffness
        self.c1 = damping
        self.m = pendulum_mass
        self.M = block_mass
        self.g = g

        self.pathway = me.LinearPathway(self.P1, self.P2)
        self.duffing_spring = me.DuffingSpring(self.k1, self.k2, self.pathway, 0)

    def to_loads(self):
        damper = me.LinearDamper(c1, self.pathway)
        damper.force

        duffing_force = self.duffing_spring.force * N.y

        loads = [(self.P1, duffing_force + damper.force * N.y), (self.P2, -self.m * self.g * N.y)]
        return loads
```

Or do we include only duffing force in the DuffingActuator class and separately obtain damping from `LinearDamper`?

Author: I've also thought of this way:
```
class DuffingActuator(me.ActuatorBase):
    def __init__(self, P1, P2, linear_stiffness, nonlinear_stiffness, damping, pendulum_mass, block_mass, g, equilibrium_length):
        self.P1 = P1
        self.P2 = P2
        self.k1 = linear_stiffness
        self.k2 = nonlinear_stiffness
        self.c1 = damping
        self.m = pendulum_mass
        self.M = block_mass
        self.g = g
        self.l = equilibrium_length

        self.pathway = me.LinearPathway(self.P1, self.P2)

    def to_loads(self):
        damper = me.LinearDamper(c1, self.pathway)

        displacement = self.pathway.length - self.l
        duffing_force = -self.k1 * displacement - self.k2 * displacement**3

        loads = [(self.P1, duffing_force * N.y + damper.force * N.y), (self.P2, -self.m * self.g * N.y)]
        return loads
```

Reviewer 2: Why use an entire class for it. You could simplify it to something like:
```python
path = me.LinearPathway(O, block_center)
spring = me.DuffingSpring(k1, k2, path)
damper = me.LinearDamper(c1, path)

loads = spring.to_loads() + damper.to_loads()
```
The gravitational force (potential energy) should not be inside the actuator. To set the potential energy you can do something like:
```python
for body in bodies:
    body.potential_energy = body.mass * g * body.masscenter.pos_from(O).dot(-N.y)
```
